internal/pkg/source/source.go

```diff
@@ -75,104 +75,185 @@ func New(in ssa.Node, config classifier) *Source {
 		marked: make(map[ssa.Node]bool),
 		config: config,
 	}
-	s.dfs(in, map[*ssa.BasicBlock]int{}, nil)
+	s.dfs(in, map[*ssa.BasicBlock]int{}, nil, false)
 	return s
 }
 
 // dfs performs Depth-First-Search on the def-use graph of the input Source.
 // While traversing the graph we also look for potential sanitizers of this Source.
 // If the Source passes through a sanitizer, dfs does not continue through that Node.
-func (s *Source) dfs(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
-	if s.marked[n] {
-		return
-	}
-	// booleans can't meaningfully be tainted
-	if isBoolean(n) {
+func (s *Source) dfs(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock, isReferrer bool) {
+	if s.shouldNotVisit(n, maxInstrReached, lastBlockVisited, isReferrer) {
 		return
 	}
+	s.preOrder = append(s.preOrder, n)
+	s.marked[n] = true
 
 	mirCopy := map[*ssa.BasicBlock]int{}
 	for m, i := range maxInstrReached {
 		mirCopy[m] = i
 	}
 
 	if instr, ok := n.(ssa.Instruction); ok {
-		// If the referrer is in a different block from the one we last visited,
-		// and it can't be reached from the block we are visiting, then stop visiting.
-		if ib := instr.Block(); lastBlockVisited != nil &&
-			ib != lastBlockVisited &&
-			!s.canReach(lastBlockVisited, ib) {
+		instrIndex, ok := indexInBlock(instr)
+		if !ok {
 			return
 		}
 
-		b := instr.Block()
-		if i, ok := indexInBlock(instr); ok && mirCopy[b] < i {
-			mirCopy[b] = i
+		if mirCopy[instr.Block()] < instrIndex {
+			mirCopy[instr.Block()] = instrIndex
 		}
-		lastBlockVisited = b
-	}
-	s.preOrder = append(s.preOrder, n)
-	s.marked[n] = true
 
-	s.visitReferrers(n, mirCopy, lastBlockVisited)
+		lastBlockVisited = instr.Block()
+	}
 
-	s.visitOperands(n, n.Operands(nil), mirCopy, lastBlockVisited)
+	s.visit(n, mirCopy, lastBlockVisited)
 }
 
-func (s *Source) visitReferrers(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
-	referrers := s.referrersToVisit(n, maxInstrReached)
+func (s *Source) shouldNotVisit(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock, isReferrer bool) bool {
+	if s.marked[n] {
+		return true
+	}
 
-	for _, r := range referrers {
-		switch v := r.(type) {
-		case *ssa.Call:
-			if callee := v.Call.StaticCallee(); callee != nil && s.config.IsSanitizer(utils.DecomposeFunction(callee)) {
-				s.sanitizers = append(s.sanitizers, &sanitizer.Sanitizer{Call: v})
-			}
+	// booleans can't meaningfully be tainted
+	if isBoolean(n) {
```

Comment:
This is a bit pathological, but if each `byte`/`rune` from a tainted `[]byte` or `[]rune` were to be sinked individually, avoiding `bytes` and `runes` would miss that. (e.g. something like `for _, r := range "secret" { fmt.Printf(r) }`; again, this is kind of pathological). A similar thing could occur with some numeric types, e.g. in theory a cryptographic key could be sinked byte-by-byte.

> we should consider excluding ... pointer types.

I think that could cause us to miss a lot of things. We don't know what a sink is going to do with its argument, e.g. dereference it.

> Conversely, if memory serves, ...

You are correct, but we also wanted to avoid things like `fmt.Println(isSource(Source{}))`, so I think avoiding all booleans is justified.

Reply:
> I think that could cause us to miss a lot of things. We don't know what a sink is going to do with a pointer argument, e.g. dereference it.

I was imprecise.  I meant within the context of basic types - specifically, the [`types.BasicKind`](https://godoc.org/go/types#BasicKind)s, which fall into:
* Invalid
* Strings
* Numeric types, including boolean
* _Unsafe_ pointers
* Untyped types
* The `Byte` and `Rune` alias for `Uint8` and `Int32` respectively.

So when I said "pointers," I should have said "unsafe pointers."  I feel like `unsafe`, like `reflect`, is out of scope for contextual analysis for us.

So of the above, I would consider taint as able to propagate to strings (typed and untyped), and optionally runes (typed and untypted) and bytes.

You could make a similar pathological argument that someone could coerce a `string -> []rune -> []int32 -> []int64` and back, but that feels like the sort of thing that is hard to avoid criticism in code review.  \shrug

Reply:
Also, this is another thing that falls into "could easily be its own PR."  I was really just stirring conversation while it was front of mind, but we should shift this conversation to in Issue to not block this PR.

Reply:
Agreed. Issue here: https://github.com/google/go-flow-levee/issues/191